### PR TITLE
gh-123961: convert `curses` to a multi-phase initialization module (PEP-489)

### PR DESCRIPTION
**Pull request by picnixz**
I think this is the very last PR for this task, conditioned to some final cleanups just for macro readability and comments.

* Issue: gh-123961

### CODE REVIEW

**Comment by picnixz** [comment]
Following our discussion on the previous PR, I also changed the `_cursesmodule` prefix to `cursesmodule` so that we have at least a bit more of consistency. I won't touch the AC-generated functions.

**Comment by picnixz** [comment]
Oh so free-threading is not happy. I'll investigate.

File: Modules/_cursesmodule.c [hunk 1]
```diff
@@ -211,16 +212,21 @@ static const char *curses_screen_encoding = NULL;
...
     if (called == TRUE) {
         return 1;
     }
-    PyErr_Format(curses_global_state.error, "must call %s() first", funcname);
+    PyObject *exc = _PyImport_GetModuleAttrString("_curses", "error");
+    if (exc != NULL) {
+        PyErr_Format(exc, "must call %s() first", funcname);
+        Py_DECREF(exc);
+    }
+    // assert(PyErr_Occurred());
```
Comment (vstinner) [inline Modules/_cursesmodule.c]:
```suggestion
    assert(PyErr_Occurred());
```

File: Modules/_cursesmodule.c [hunk 2]
```diff
@@ -211,16 +212,21 @@ static const char *curses_screen_encoding = NULL;
...
+ * is imported on demand.
  */
 static inline int
 _PyCursesCheckFunction(int called, const char *funcname)
 {
     if (called == TRUE) {
         return 1;
     }
-    PyErr_Format(curses_global_state.error, "must call %s() first", funcname);
+    PyObject *exc = _PyImport_GetModuleAttrString("_curses", "error");
```
Comment (vstinner) [inline Modules/_cursesmodule.c:227]: If the extension is loaded more than once, the caller may get the wrong "error" exception.

Is there a way to disallow loading a C extension more than once? Anyway, I don't think that it's a good idea that two _curses extension control the same terminal.

Currently, it's possible to have two instances:

```
$ ./python
>>> import sys
>>> import _curses as curses1
>>> del sys.modules['_curses']
>>> import _curses as curses2
>>> curses2 is curses1
False
>>> curses2.error is curses1.error
False
```

Comment (picnixz) [reply Modules/_cursesmodule.c:227]: > Is there a way to disallow loading a C extension more than once

The docs suggest having a global flag (see https://docs.python.org/3/howto/isolating-extensions.html#opt-out-limiting-to-one-module-object-per-process). We also say "The goal is to make extension modules safe at the C level, not to make hacks behave intuitively. Mutating sys.modules “manually” counts as a hack." so I don't know if we should care (https://docs.python.org/3/howto/isolating-extensions.html#surprising-edge-cases).

Comment (picnixz) [reply Modules/_cursesmodule.c:227]: Ah sorry, the global flag is for multiple interpreters not for reloading. For reloading modules I think we should have the datetime module approach (I think there is something out there that takes care of reloading but I don't whether it'd help)

Comment (vstinner) [reply Modules/_cursesmodule.c:227]: > The docs suggest having a global flag

I would be fine with a global flag.

Comment (picnixz) [reply Modules/_cursesmodule.c:227]: ~~Ok, I found a way to have the same error type. Actually I shouldn't use `PyDict_SetItemString` but I can just use `PyModule_AddType`. I don't know why it works though but I copied what was being done in `sqlite3`.~~

**Review by vstinner — COMMENTED** [review]
Can you try to add a global variable to prevent having more than 1 instance at the same time?

**Comment by picnixz** [comment]
Here's the output:

```pycon
>>> import sys
...
>>> import _curses as curses1
>>> del sys.modules['_curses']
...
>>> import _curses as curses2
...
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    import _curses as curses2
ImportError: module 'curses' can only be loaded once per process
```

**Comment by vstinner** [comment]
`cursesmodule_free()` should set `curses_module_loaded` to 0. So you can load-unload the extension multiple times:

```py
import sys
import gc

print("load-unload")
import _curses
del _curses
del sys.modules['_curses']
gc.collect()

print("load again")
import _curses
del _curses
del sys.modules['_curses']
```

File: Modules/_cursesmodule.c [hunk 3]
```diff
@@ -5181,33 +5221,27 @@ cursesmodule_exec(PyObject *module)
...
+    {Py_mod_gil, Py_MOD_GIL_NOT_USED},
+    {0, NULL}
+};
+
+static struct PyModuleDef cursesmodule = {
     PyModuleDef_HEAD_INIT,
     .m_name = "_curses",
-    .m_size = -1,
-    .m_methods = PyCurses_methods,
+    .m_doc = NULL,
```
Comment (vstinner) [inline Modules/_cursesmodule.c]: You can remove it.

**Comment by vstinner** [comment]
> The cursesmodule_free() does not seem to be called even if gc.collect() is called (at least not in the REPL).

It works if you run my [script](https://github.com/python/cpython/pull/124965#issuecomment-2397184718):

```
load-unload
load again
```

**Comment by picnixz** [comment]
After some tests, it appears that the REPL only calls `free()` upon exiting the process. So, reloading curses inside the REPL won't work. In a normal script, however:

```sh
$ read -r -d '' code << EOM
import sys

print("load-unload")
import _curses
del _curses
del sys.modules['_curses']

print("load again")
import _curses
del _curses
del sys.modules['_curses']
EOM
$ ./python -c "$code"
load-unload
load again
Traceback (most recent call last):
  File "<string>", line 9, in <module>
    import _curses
ImportError: module 'curses' can only be loaded once per process
```

When `free` is called via `gc.collect()`:

```sh
$ read -r -d '' code << EOM
import sys
import gc

print("load-unload")
import _curses
del _curses
del sys.modules['_curses']
gc.collect()

print("load again")
import _curses
del _curses
del sys.modules['_curses']
EOM
$ ./python -c "$code"
load-unload
load again
```

**Comment by vstinner** [comment]
Merged, thanks. That's a nice step forward for the _curses extension :-)

**Comment by vstinner** [comment]
Python no longer leaks at exit:

```
$ ./python -X showrefcount -c 'import _curses'
[0 refs, 0 blocks]
```

Python 3.13 for comparison:

```
$ ./python -X showrefcount -c 'import _curses'
[234 refs, 159 blocks]
```

**Pull request merged**